File: java/0015.3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums){
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        for(int i = 0; i < nums.length - 2; ++i){
            if(nums[i] > 0) break; // 如果第一个数大于0，那就肯定没有解
            if(i > 0 && nums[i] == nums[i-1]) continue;
            int L = i + 1,R = nums.length-1;
            int target = -nums[i];
            while(L < R){
                int sum = nums[L] + nums[R];
                if(sum == target){
                    while(L < R && nums[L] == nums[L+1]) ++L;
                    while(L < R && nums[R] == nums[R-1]) --R;
                    ArrayList<Integer> list = new ArrayList<Integer>();
                    list.add(nums[i]);
                    list.add(nums[L]);
                    list.add(nums[R]);
                    result.add(list);
                    ++L;
                    --R;
                }
                else if(sum < target) ++L;
                else --R;
            }
        }
        return result;
    }
}

File: python/0015. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []

        for first in range(n):
            if nums[first] > 0:
                break
            if first>0 and nums[first] ==nums[first-1]:
                continue
            
            third = n-1
            target = -nums[first]

            second=first+1
            while second<third:
                if nums[second]+nums[third]<target:
                    second+=1
                elif nums[second]+nums[third]>target:
                    third-=1
                else:
                    ans.append([nums[first],nums[second],nums[third]])
                    while second<third and nums[second]==nums[second+1]:
                        second+=1
                    while second<third and nums[third]==nums[third-1]:
                        third-=1
                    second+=1
                    third-=1
        return ans
